lib/dice/src/cert.rs

```diff
@@ -0,0 +1,340 @@
+);
+
+impl DeviceIdCert {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        let mut buf = [0u8; deviceid_cert_tmpl::SIZE];
+        deviceid_cert_tmpl::fill(&mut buf);
+
+        Self(buf)
+    }
+
+    pub fn from(bytes: &[u8; deviceid_cert_tmpl::SIZE]) -> Self {
+        Self(*bytes)
+    }
+
+    pub fn as_bytes(&self) -> &[u8; deviceid_cert_tmpl::SIZE] {
+        &self.0
+    }
+
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn set_serial_number(mut self, sn: u8) -> Self {
+        self.0[deviceid_cert_tmpl::SERIAL_NUMBER_START
+            ..deviceid_cert_tmpl::SERIAL_NUMBER_END]
+            .copy_from_slice(&sn.to_be_bytes());
+
+        self
+    }
+
+    pub fn set_issuer_sn(mut self, sn: &SerialNumber) -> Self {
+        self.0[deviceid_cert_tmpl::ISSUER_SN_START
+            ..deviceid_cert_tmpl::ISSUER_SN_END]
+            .copy_from_slice(sn.as_bytes());
+
+        self
+    }
+
+    pub fn set_notbefore(
+        mut self,
+        utctime: &[u8; deviceid_cert_tmpl::NOTBEFORE_LENGTH],
+    ) -> Self {
+        self.0[deviceid_cert_tmpl::NOTBEFORE_START
+            ..deviceid_cert_tmpl::NOTBEFORE_END]
+            .copy_from_slice(utctime);
+
+        self
+    }
+
+    pub fn set_subject_sn(mut self, sn: &SerialNumber) -> Self {
+        self.0[deviceid_cert_tmpl::SUBJECT_SN_START
+            ..deviceid_cert_tmpl::SUBJECT_SN_END]
+            .copy_from_slice(sn.as_bytes());
+
+        self
+    }
+
+    pub fn set_pub(
+        mut self,
+        pubkey: &[u8; PUBLICKEY_SERIALIZED_LENGTH],
+    ) -> Self {
+        self.0[deviceid_cert_tmpl::PUB_START..deviceid_cert_tmpl::PUB_END]
+            .copy_from_slice(pubkey);
+
+        self
+    }
+
+    pub fn get_pub(&self) -> &[u8; deviceid_cert_tmpl::PUB_LEN] {
+        self.0[deviceid_cert_tmpl::PUB_START..deviceid_cert_tmpl::PUB_END]
+            .try_into()
+            .unwrap_lite()
+    }
+
+    pub fn set_sig(mut self, sig: &[u8; SIGNATURE_SERIALIZED_LENGTH]) -> Self {
+        self.0[deviceid_cert_tmpl::SIG_START..deviceid_cert_tmpl::SIG_END]
+            .copy_from_slice(sig);
+
+        self
+    }
+
+    pub fn get_sig(&self) -> &[u8; deviceid_cert_tmpl::SIG_LEN] {
+        self.0[deviceid_cert_tmpl::SIG_START..deviceid_cert_tmpl::SIG_END]
+            .try_into()
+            .unwrap_lite()
+    }
+
+    pub fn get_signdata(&self) -> &[u8; deviceid_cert_tmpl::SIGNDATA_LENGTH] {
+        self.0[deviceid_cert_tmpl::SIGNDATA_START
+            ..deviceid_cert_tmpl::SIGNDATA_END]
+            .try_into()
+            .unwrap_lite()
+    }
+
+    pub fn sign(self, keypair: &Keypair) -> Self {
+        let signdata = self.get_signdata();
+        let sig = keypair.sign(signdata);
+
+        self.set_sig(&sig.to_bytes())
+    }
+}
+
+#[derive(Deserialize, Serialize, SerializedSize)]
+pub struct AliasCert(#[serde(with = "BigArray")] [u8; alias_cert_tmpl::SIZE]);
```

Comment:
I wrestled with this quite a bit. The issue I ran into is that each function that's duplicated across the two types is operating on different offsets. Using `set_serial_number` as an example the impl for DeviceIdCert is:

```
pub fn set_serial_number(mut self, sn: u8) -> Self {
    self.0[deviceid_cert_tmpl::SERIAL_NUMBER_START
        ..deviceid_cert_tmpl::SERIAL_NUMBER_END]
        .copy_from_slice(&sn.to_be_bytes());

        self
}
```
while the impl for AliasCert is:
```
pub fn set_serial_number(mut self, sn: u8) -> Self {
    self.0[alias_cert_tmpl::SERIAL_NUMBER_START
        ..alias_cert_tmpl::SERIAL_NUMBER_END]
        .copy_from_slice(&sn.to_be_bytes());

    self
}
```

The only difference between these two functions are the different offsets used to create the slices. I may be understanding the Rust part of this wrong but AFAIK default implementations in traits cannot assume anything about the fields in the types so to get these offsets we'd have to define `get_*` functions in the trait for each offset. This will double the number of boilerplate functions (a `get` function for each start and end offset) though this may be an improvement since these `get` functions will only return an offset, and so as far as boilerplate goes that's pretty good.

I think after all this I've just talked myself into the idea. I'll hack something up to see how it looks.

Reply:
This is actually better than I had expected. I didn't realize how useful associated constants in traits would be here.

Reply:
I ended up banging my head against a known issue in my attempts to create a more generic interface with shared implementation: https://github.com/rust-lang/rust/issues/60551. I'm currently trying some of the suggested workarounds that primarily rely on using constant generic parameters.

Reply:
This wasn't as bad as expected. All of the implementation from the DeviceIdCert is now in a trait. The only additional code in the AliasCert is a get/set for the FWID (the only field in the AliasCert that's not in the DeviceIdCert). I had to compromise on a few sized arrays in the interface that are now slices but these can be added back once https://github.com/rust-lang/rust/issues/60551 moves forward.